File: python/TemplateMethod/beverage.py
from abc import abstractmethod


class CaffeineBeverageWithHook:
    def __init__(self) -> None:
        pass

    def prepareRecipe(self):
        self.boilWater()
        self.brew()
        self.PourInCup()
        if self.CustomerWantsCondiments():
            self.addCondiments()

    def boilWater(self):
        print("Boiling water")

    def PourInCup(self):
        print("Pouring into cup")

    def CustomerWantsCondiments(self) -> bool:
        return True

    @abstractmethod
    def brew(self):
        pass

    @abstractmethod
    def addCondiments(self):
        pass


class CoffeeWithHook(CaffeineBeverageWithHook):
    def __init__(self) -> None:
        super().__init__()

    def brew(self):
        print("Dripping Coffee through filter")

    def addCondiments(self):
        print("Adding Sugar and Milk")

    def CustomerWantsCondiments(self) -> bool:
        answer = self.getUserInput()
        if answer.lower().startswith("y"):
            return True
        else:
            return False

    def getUserInput(self) -> str:
        print("Would you like sugar and milk with your coffee (y/n)? ", end="")
        answer = input()
        if not answer:
            return "no"
        return answer


class TeaWithHook(CaffeineBeverageWithHook):
    def __init__(self) -> None:
        super().__init__()

    def brew(self):
        print("Steeping the tea")

    def addCondiments(self):
        print("Adding Lemon")

    def CustomerWantsCondiments(self) -> bool:
        answer = self.getUserInput()
        if answer.lower().startswith("y"):
            return True
        else:
            return False

    def getUserInput(self) -> str:
        print("Would you like lemon with your tea (y/n)? ", end="")
        answer = input()
        if not answer:
            return "no"
        return answer

File: c/TemplateMethod/beverage.c
#include "beverage.h"
void CaffeineBeverageWithHookNew(CaffeineBeverageWithHook *self)
{
    self->prepareRecipe = CBWH_prepareRecipe;
    self->boilWater = CBWH_boilWater;
    self->PourInCup = CBWH_PourInCup;
    self->CustomerWantsCondiments = CBWH_CustomerWantsCondiments;
}
void CBWH_prepareRecipe(CaffeineBeverageWithHook *self)
{
    self->boilWater(self);
    self->brew(self);
    self->PourInCup(self);
    if (self->CustomerWantsCondiments(self))
    {
        self->addCondiments(self);
    }
}
void CBWH_boilWater(CaffeineBeverageWithHook *self)
{
    printf("Boiling water\n");
}
void CBWH_PourInCup(CaffeineBeverageWithHook *self)
{
    printf("Pouring into cup\n");
}
bool CBWH_CustomerWantsCondiments(CaffeineBeverageWithHook *self)
{
    return true;
}
void CoffeeWithHookNew(CoffeeWithHook *self)
{
    self->prepareRecipe = CoffeeWH_prepareRecipe;
    self->boilWater = CoffeeWH_boilWater;
    self->PourInCup = CoffeeWH_PourInCup;
    self->CustomerWantsCondiments = CoffeeWH_CustomerWantsCondiments;
    self->brew = CoffeeWH_brew;
    self->addCondiments = CoffeeWH_addCondiments;
    self->getUserInput = CoffeeWH_getUserInput;
}
void CoffeeWH_prepareRecipe(CoffeeWithHook *self)
{
    CBWH_prepareRecipe((CaffeineBeverageWithHook *)self);
}
void CoffeeWH_boilWater(CoffeeWithHook *self)
{
    CBWH_boilWater((CaffeineBeverageWithHook *)self);
}
void CoffeeWH_PourInCup(CoffeeWithHook *self)
{
    CBWH_PourInCup((CaffeineBeverageWithHook *)self);
}
bool CoffeeWH_CustomerWantsCondiments(CoffeeWithHook *self)
{
    char c = self->getUserInput(self);
    if (c == 'y' || c == 'Y')
        return true;
    else
        return false;
}
void CoffeeWH_brew(CoffeeWithHook *self)
{
    printf("Dripping Coffee through filter\n");
}
void CoffeeWH_addCondiments(CoffeeWithHook *self)
{
    printf("Adding Sugar and Milk\n");
}
char CoffeeWH_getUserInput(CoffeeWithHook *self)
{
    char s[20];
    printf("Would you like sugar and milk with your coffee (y/n)? ");
    gets(s);
    return s[0];
}

void TeaWithHookNew(TeaWithHook *self)
{
    self->prepareRecipe = TeaWH_prepareRecipe;
    self->boilWater = TeaWH_boilWater;
    self->PourInCup = TeaWH_PourInCup;
    self->CustomerWantsCondiments = TeaWH_CustomerWantsCondiments;
    self->brew = TeaWH_brew;
    self->addCondiments = TeaWH_addCondiments;
    self->getUserInput = TeaWH_getUserInput;
}

void TeaWH_prepareRecipe(TeaWithHook *self)
{
    CBWH_prepareRecipe((CaffeineBeverageWithHook *)self);
}
void TeaWH_boilWater(TeaWithHook *self)
{
    CBWH_boilWater((CaffeineBeverageWithHook *)self);
}
void TeaWH_PourInCup(TeaWithHook *self)
{
    CBWH_PourInCup((CaffeineBeverageWithHook *)self);
}
bool TeaWH_CustomerWantsCondiments(TeaWithHook *self)
{
    char c = self->getUserInput(self);
    if (c == 'y' || c == 'Y')
        return true;
    else
        return false;
}
void TeaWH_brew(TeaWithHook *self)
{
    printf("Steeping the tea\n");
}
void TeaWH_addCondiments(TeaWithHook *self)
{
    printf("Adding Lemon\n");
}
char TeaWH_getUserInput(TeaWithHook *self)
{
    char s[20];
    printf("Would you like lemon with your coffee (y/n)? ");
    gets(s);
    return s[0];
}